exetera/core/abstract_types.py

```diff
@@ -133,6 +133,112 @@ def __len__(self):
         raise NotImplementedError()
 
 
+class DataFrame(ABC):
+    """
+    DataFrame is a table of data that contains a list of Fields (columns)
+    """
+
+    @abstractmethod
+    def add(self):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def create_group(self, name):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def create_numeric(self, name, nformat, timestamp=None, chunksize=None):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def create_indexed_string(self, name, timestamp=None, chunksize=None):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def create_fixed_string(self, name, length, timestamp=None, chunksize=None):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def create_categorical(self, name, nformat, key, timestamp=None, chunksize=None):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def create_timestamp(self, name, timestamp=None, chunksize=None):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def __contains__(self, name):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def contains_field(self, field):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def __getitem__(self, name):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_field(self, name):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_name(self, field):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def __setitem__(self, name, field):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def __delitem__(self, name):
+        raise NotImplementedError()
+
+    @abstractmethod
+    def delete_field(self, field):
+        raise NotImplementedError()
+
+    @abstractmethod

```

Comment:
I think we should lose the list method. Apart from anything else, it actually returns a tuple! I also just don't think it is necessary and it isn't part of any existing interface that we are attempting to mimic. I know that Session has a list_datasets but it probably should be tweaked to have keys, values and items as well as __getitem__.